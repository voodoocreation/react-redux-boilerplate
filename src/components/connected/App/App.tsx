import { NextContext } from "next";
import withReduxSaga from "next-redux-saga";
import withRedux from "next-redux-wrapper";
import { DefaultQuery } from "next-server/router";
import NextApp, {
  AppProps,
  Container,
  DefaultAppIProps,
  NextAppContext
} from "next/app";
import * as React from "react";
import { addLocaleData } from "react-intl";
import { IntlProvider } from "react-intl-redux";
import { Provider } from "react-redux";

import routes from "../../../../next.routes";
import * as actions from "../../../actions/root.actions";
import { isServer } from "../../../helpers/dom";
import { createStore, TStore } from "../../../store/root.store";
import Page from "../Page/Page";

import en from "react-intl/locale-data/en";

addLocaleData([...en]);

export type TContext<Q extends DefaultQuery> = NextContext<Q> & {
  store: TStore;
};

export type TAppContext<Q extends DefaultQuery> = NextAppContext & {
  ctx: TContext<Q>;
  store: TStore;
};

interface IProps extends DefaultAppIProps, AppProps {
  intlProps: {
    locale: string;
    initialNow: Date;
  };
  store: TStore;
}

const getIntlProps = (ctx: NextContext) => {
  const requestProps = isServer()
    ? ctx.req
    : window.__NEXT_DATA__.props.initialProps.intlProps;
  const { locale } = requestProps;

  return {
    initialNow: Date.now(),
    locale: locale || "en-NZ"
  };
};

class App<P extends IProps> extends NextApp<P> {
  public static async getInitialProps({ ctx, Component }: TAppContext<any>) {
    let pageProps = {};

    const unsubscribe = ctx.store.subscribe(() => {
      const state = ctx.store.getState();

      if (state.app.error && ctx.res) {
        ctx.res.statusCode = state.app.error.status;
        unsubscribe();
      }
    });

    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx);
    }

    const intlProps = getIntlProps(ctx);

    return { pageProps, intlProps };
  }

  constructor(props: P) {
    super(props);

    routes.Router.onRouteChangeStart = this.onRouteChangeStart;
    routes.Router.onRouteChangeComplete = this.onRouteChangeComplete;
    routes.Router.onRouteChangeError = this.onRouteChangeError;

    props.store.dispatch(
      actions.initApp.started({
        locale: props.intlProps.locale
      })
    );
  }

  public componentDidMount() {
    const { store } = this.props;

    store.dispatch(actions.setCurrentRoute(routes.Router.route));
  }

  public render() {
    const { Component, intlProps, pageProps, store } = this.props;

    return (
      <Container>
        <Provider store={store}>
          <IntlProvider
            defaultLocale="en-NZ"
            initialNow={intlProps.initialNow}
            textComponent={React.Fragment}
          >
            <Page>
              <Component {...pageProps} />
            </Page>
          </IntlProvider>
        </Provider>
      </Container>
    );
  }

  private onRouteChangeStart = (path: string) => {
    const { store } = this.props;

    store.dispatch(actions.changeRoute.started(path));
  };

  private onRouteChangeComplete = (path: string) => {
    const { store } = this.props;

    store.dispatch(actions.changeRoute.done({ params: path, result: {} }));
  };

  private onRouteChangeError = (error: Error, path: string) => {
    const { store } = this.props;

    store.dispatch(
      actions.changeRoute.failed({
        error: {
          message: error.toString(),
          status: 500
        },
        params: path
      })
    );
  };
}

const AppWrapped = withRedux(createStore)(withReduxSaga(App));

export default AppWrapped;
