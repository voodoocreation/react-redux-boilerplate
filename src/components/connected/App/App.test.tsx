import { mount } from "enzyme";
import merge from "lodash.merge";
import * as React from "react";

import App from "./App";

jest.mock("../../../../next.routes", () => ({
  Router: {
    route: ""
  }
}));

import routes from "../../../../next.routes";

const g: any = global;

const setup = async (fn: any, fromTestProps?: any, isServer = false) => {
  jest.clearAllMocks();

  g.isServer = isServer;
  g.__NEXT_DATA__ = {
    props: {
      initialProps: {
        intlProps: {}
      }
    }
  };

  const Component = () => <div className="PageComponent" />;
  const appProps = merge(
    {
      Component,
      asPath: "",
      ctx: {
        isServer: false,
        pathname: "",
        query: {},
        req: {
          intlMessages: {},
          locale: "en-NZ"
        },
        res: {}
      },
      router: {
        pathname: ""
      }
    },
    fromTestProps
  );
  const initialProps = await App.getInitialProps(appProps);
  const props = {
    ...appProps,
    ...initialProps
  };

  return {
    actual: fn(<App {...props} />),
    props
  };
};

describe("[connected] <App />", () => {
  const addEventListener = g.addEventListener;
  const removeEventListener = g.removeEventListener;

  beforeAll(() => {
    g.addEventListener = jest.fn((...args) => addEventListener(...args));
    g.removeEventListener = jest.fn((...args) => removeEventListener(...args));
  });

  it("mounts application correctly on the server", async () => {
    const { actual } = await setup(
      mount,
      {
        ctx: { isServer: g.isServer }
      },
      true
    );

    expect(actual.render()).toMatchSnapshot();
    actual.unmount();
  });

  it("mounts application correctly on the client", async () => {
    const { actual } = await setup(mount);

    expect(actual.render()).toMatchSnapshot();
    actual.unmount();
  });

  it("gets `initialProps` from component correctly", async () => {
    const test = "Test";
    const Component: any = () => <div className="PageComponent" />;

    Component.getInitialProps = async () => ({ test });

    const { actual, props } = await setup(mount, { Component });

    expect(props.initialProps.pageProps).toEqual({ test });

    actual.unmount();
  });

  describe("when navigating within the app", () => {
    let result: any;

    it("mounts the component", async () => {
      result = await setup(mount);
    });

    it("handles the onRouteChangeStart event correctly", async () => {
      routes.Router.onRouteChangeStart("/");

      expect(result.props.ctx.store.getState().app.transitioningTo).toBe("/");
    });

    it("handles the onRouteChangeComplete event correctly", async () => {
      routes.Router.onRouteChangeComplete("/");

      expect(
        result.props.ctx.store.getState().app.transitioningTo
      ).toBeUndefined();
      expect(result.props.ctx.store.getState().app.currentRoute).toBe("/");
    });

    it("handles the onRouteChangeError event correctly", async () => {
      routes.Router.onRouteChangeError(new Error("Server error"), "/");

      expect(
        result.props.ctx.store.getState().app.transitioningTo
      ).toBeUndefined();
      expect(result.props.ctx.store.getState().app.error).toEqual({
        message: "Error: Server error",
        status: 500
      });
    });

    it("unmounts the component", () => {
      result.actual.unmount();
    });
  });
});
