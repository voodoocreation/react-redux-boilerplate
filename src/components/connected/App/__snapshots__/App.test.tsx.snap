// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[connected] <App /> mounts application correctly on the client 1`] = `
<article
  class="Page"
>
  <main
    class="Page--body container"
    role="main"
  >
    <div
      class="PageComponent"
    />
  </main>
</article>
`;

exports[`[connected] <App /> mounts application correctly on the server 1`] = `
<article
  class="Page"
>
  <main
    class="Page--body container"
    role="main"
  >
    <div
      class="PageComponent"
    />
  </main>
</article>
`;
